From aaf1a25c65647aec4c6fd583582d324392bff1ef Mon Sep 17 00:00:00 2001
From: Charles Bousseau <cbousseau@anaconda.com>
Date: Sun, 20 Jul 2025 14:03:26 -0400
Subject: [PATCH] metal gpu selection

In macOS, in order for the system to provide a default Metal device object, you must link to the Core Graphics framework. 
You usually need to do this explicitly if you’re writing apps that don’t use graphics by default, such as command line tools.
https://developer.apple.com/documentation/metal/1433401-mtlcreatesystemdefaultdevice?language=objc
Systems with Apple silicon only have one GPU, which removes the need to choose a GPU.
https://developer.apple.com/documentation/metal/mtldevice/1433409-lowpower#discussion

I did try linking to CoreGraphics, but MTLCreateSystemDefaultDevice was still returning nil.
---
 ggml/src/ggml-metal/ggml-metal.m | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/ggml/src/ggml-metal/ggml-metal.m b/ggml/src/ggml-metal/ggml-metal.m
index dc391a0d4..2083e2a31 100644
--- a/ggml/src/ggml-metal/ggml-metal.m
+++ b/ggml/src/ggml-metal/ggml-metal.m
@@ -92,6 +92,25 @@
 
     if (ctx->mtl_device == nil) {
         ctx->mtl_device = MTLCreateSystemDefaultDevice();
+        if (ctx->mtl_device == nil) {
+          /*
+            In macOS, in order for the system to provide a default Metal device object, you must link to the Core Graphics framework.
+            You usually need to do this explicitly if you're writing apps that don't use graphics by default, such as command line tools.
+            > https://developer.apple.com/documentation/metal/1433401-mtlcreatesystemdefaultdevice?language=objc
+            Systems with Apple silicon only have one GPU, which removes the need to choose a GPU.
+            > https://developer.apple.com/documentation/metal/mtldevice/1433409-lowpower#discussion
+           */
+            NSArray * devices = MTLCopyAllDevices();
+            for (id<MTLDevice> dev in devices) {
+                if (dev != nil) {
+                    if (ctx->mtl_device == nil) {
+                        ctx->mtl_device = dev;
+                    } else {
+                        [dev release];
+                    }
+                }
+            }
+        }
 
         ctx->has_simdgroup_reduction  = [ctx->mtl_device supportsFamily:MTLGPUFamilyApple7];
         ctx->has_simdgroup_reduction |= [ctx->mtl_device supportsFamily:MTLGPUFamilyMetal3_GGML];
-- 
2.39.5 (Apple Git-154)

